crates/iota-indexer/src/indexer_reader.rs

```diff
@@ -225,6 +225,116 @@ impl IndexerReader {
         Ok(stored_object)
     }
 
+    /// Fetches the past object read by object ID and version.
+    /// - If `before_version` is `false`, it looks for the exact version.
+    /// - If `true`, it finds the object read for the latest version before the
+    ///   given one.
+    pub(crate) async fn get_past_object_read_in_blocking_task(
+        &self,
+        object_id: ObjectID,
+        version: SequenceNumber,
+        before_version: bool,
+    ) -> Result<PastObjectRead, IndexerError> {
+        let (history_object, latest_version) = self
+            .spawn_blocking(move |this| this.query_past_object(&object_id, version, before_version))
+            .await?;
+
+        match (history_object, latest_version) {
+            (Some(object), _) => {
+                object
+                    .try_into_past_object_read(self.package_resolver.clone())
+                    .await
+            }
+
+            // If the object never existed
+            (None, None) => Ok(PastObjectRead::ObjectNotExists(object_id)),
+
+            // If the object exists, but the requested version is too high
+            (None, Some(latest)) if version.value() > latest as u64 => {
+                Ok(PastObjectRead::VersionTooHigh {
+                    object_id,
+                    asked_version: version,
+                    latest_version: SequenceNumber::from(latest as u64),
+                })
+            }
+
+            // If the object exists but not at this version
+            (None, Some(_)) => Ok(PastObjectRead::VersionNotFound(object_id, version)),
+        }
+    }
+
+    /// Fetches a past object by its ID and version.
+    ///
+    /// - If `before_version` is `false`, it looks for the exact version.
+    /// - If `true`, it finds the latest version before the given one.
+    ///
+    /// Also retrieves the latest available version of the object to check if:
+    /// - The requested version is too high.
+    /// - The object exists but not at this version.
+    /// - The object doesn’t exist at all.
+    ///
+    /// Searches `objects_history`.
+    fn query_past_object(
+        &self,
+        object_id: &ObjectID,
+        version: SequenceNumber,
+        before_version: bool,
+    ) -> Result<(Option<StoredHistoryObject>, Option<i64>), IndexerError> {
```

Comment:
Nit: I think it would make the function more readable and intuitive if you would return directly `PostObjectRead` here instead of the tuple. There are very general semantics in `Option<i64>` and the signature implies that all combinations are possible (e.g. `Some(object), None`), although it wouldn't make sense.

Reply:
Hey @kodemartin, good point and was also thinking about this. Refactored: https://github.com/iotaledger/iota/pull/5865/commits/0f664bfdf02c66f8091f661483995fd9558b180b